txscript/pkscript.go

```diff
@@ -0,0 +1,194 @@
+package txscript
+
+import (
+	"crypto/sha256"
+	"errors"
+	"fmt"
+
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+	"golang.org/x/crypto/ripemd160"
+)
+
+const (
+	// pubKeyHashSigScriptLen is the length of a signature script attempting
+	// to spend a P2PKH script. This length is determined by the following:
+	//   0x48 (72 byte data push) | <71 byte sig> | <1 byte sighash flag> |
+	//   0x21 (33 byte data push) | <33 byte compressed pubkey>
+	pubKeyHashSigScriptLen = 107
+
+	// witnessV0PubKeyHashLen is the length of a P2WPKH script.
+	witnessV0PubKeyHashLen = 22
+
+	// scriptHashLen is the length of a P2SH script.
+	scriptHashLen = 23
+
+	// witnessV0ScriptHashLen is the length of a P2WSH script.
+	witnessV0ScriptHashLen = 34
+
+	// maxLen is the maximum script length supported by ParsePkScript.
+	maxLen = witnessV0ScriptHashLen
+)
+
+var (
+	// ErrUnsupportedScriptType ...
+	ErrUnsupportedScriptType = errors.New("unsupported script type")
+)
+
+// PkScript is a wrapper struct around a byte array, allowing it to be used
+// as a map index.
+type PkScript struct {
+	// class is the type of the script encoded within the byte array. This
+	// is used to determine the correct length of the script within the byte
+	// array.
+	class ScriptClass
+
+	// script is the script contained within a byte array. If the script is
+	// smaller than the length of the byte array, it will be padded with 0s
+	// at the end.
+	script [maxLen]byte
+}
+
+// ParsePkScript parses an output script into the PkScript struct.
+// ErrUnsupportedScriptType is returned when attempting to parse an unsupported
+// script type.
+func ParsePkScript(pkScript []byte) (PkScript, error) {
+	var outputScript PkScript
+	scriptClass, _, _, err := ExtractPkScriptAddrs(
+		pkScript, &chaincfg.MainNetParams,
+	)
+	if err != nil {
+		return outputScript, fmt.Errorf("unable to parse script type: "+
+			"%v", err)
+	}
+
+	if !isSupportedScriptType(scriptClass) {
+		return outputScript, ErrUnsupportedScriptType
+	}
+
+	outputScript.class = scriptClass
+	copy(outputScript.script[:], pkScript)
+
+	return outputScript, nil
+}
+
+// isSupportedScriptType determines whether the script type is supported by the
+// PkScript struct.
+func isSupportedScriptType(class ScriptClass) bool {
+	switch class {
+	case WitnessV0PubKeyHashTy, ScriptHashTy, WitnessV0ScriptHashTy:
+		return true
+	default:
+		return false
+	}
+}
+
+// Class returns the script type.
+func (s PkScript) Class() ScriptClass {
+	return s.class
+}
+
+// Script returns the script as a byte slice without any padding.
+func (s PkScript) Script() []byte {
```

Comment:
Perhaps we should make a copy instead, so the callers can't accidentally mutate our underlying script storage.

Reply:
Fixed.